New ldap state module and accompanying execution module

rhansen opened the pull request:
This is a rough draft of a new `ldap` state module.  I also created a new `ldap2` execution module to support the state module.

There are a number of issues with these two new modules:
- I don't like the name `ldap2` for the execution module because it sounds like it's for the old LDAP protocol version 2.  The name `ldap` is already taken by the `ldapmod` execution module.  (I didn't use the existing `ldapmod` module because I don't like `search()`'s API.)
- I'm not sure the API is the best.  In particular, the depth of some of the structures is pretty extreme (lists of dicts containing lists of dicts containing lists).
- The code is not well tested (e.g., I haven't tested error conditions like permission denied due to ACL rules, and I have no idea how it will handle non-ASCII text).
- There are no test cases for the `ldap2` execution module.
- The unit test code for the `ldap` state module is gross.  Really gross.
- There aren't enough test cases for the `ldap` state module.
- I'm sure there's lots of lint issues.
- There should be more examples in the documentation, especially in the `ldap2` execution module function documentation.
- Other stuff I'm forgetting.

I would love some feedback, but would especially love patches.  I don't have much time to continue working on these modules, especially since they work well enough to solve the problem I wrote them to solve.

Addresses #10076, #25911

rhansen committed:
ignore more virtualenv directories

rhansen committed:
new ldap3 execution module

This is an alternative to the existing ldapmod module.

rhansen committed:
new ldap state module

richardhansen commented:
Oops, I changed the return value and forgot to update the tests.  I'll fix that.

jfindlay commented:
@richardhansen, this is excellent, thanks.  There are some [lint errors](https://jenkins.saltstack.com/job/salt-pr-lint-n/9063/violations/) and some [other tests](https://jenkins.saltstack.com/job/salt-pr-rs-cent7-n/7892/#showFailuresLink) that need to be updated (you can ignore the git test failures).

richardhansen commented:
I pushed a reroll.  I didn't fix one of the lint issues because I'll have to think more about the right way to address it.

s0undt3ch commented:
Since this is a preparation to possibly deprecate the old ldap module, we could use this rewrite to use the ldap3 library which supports python 3. The ldap lib used here doesn't, at least, the last time I checked it.

... this is just a suggestion ...

richardhansen commented:
> Since this is a preparation to possibly deprecate the old ldap module, we could use this rewrite to use the ldap3 library which supports python 3.

I admit that I wasn't aware of the [`ldap3`](https://github.com/cannatag/ldap3) Python module.  Ubuntu 14.04 LTS doesn't have a package for it.

I skimmed the documentation and it looks like it wouldn't be too difficult to modify the code to support either module (try importing one and if that fails try importing the other).  I see two challenges in adding support for `ldap3`:
  * The `connect_spec` parameter in the proposed `ldap3` execution module has options specific to the `ldap` module.  The API will have to be adjusted to work for either module.
  * I'm not sure how the connection open/close semantics differ.  We may want to put the connection object in a wrapper that can act as a context manager and use it as such.

I did some searching and discovered that there is also:
  * [`ldaptor`](https://github.com/twisted/ldaptor) which appears to only support Python 2, and
  * [`tldap`](https://github.com/Karaage-Cluster/python-tldap) which appears to support both Python 2 and 3.  It's very Django-centric, so maybe it's not a good choice (I don't know Django at all so I can't say)

I don't know how mature, stable, and available any of these modules are, or how robust of a development community they have.  The `ldap3` documentation has lots of errors (mostly typos, but some substantive), which makes me worried that there isn't a large community of users and developers.  You mentioned `ldap3` specifically; is there a reason you suggested that over `tldap`?

> The ldap lib used here doesn't, at least, the last time I checked it.

Correct, the Python `ldap` module only works with Python 2.

> 
> ... this is just a suggestion ...

I like the idea.  I'm just not sure when I'll have the time to add support for `ldap3`.  Perhaps I can tweak the API a bit to work with either module, then later I (or someone else) can quietly add `ldap3` support without breaking compatibility with users' existing setups.

s0undt3ch commented:
I've done some work with ldap3, hence me bringing it into the issue. I basically needed python3 support when I needed it.
ldap3 connection object is itself a context manager, so, as long as it's used in a context manager manner, it takes care of opening and closing the connection. It also has async support and some other goodies. It's also full python, no binary bindings need to be compiled.

Regarding your suggestions, well:
* ldaptor uses twisted; I've used twisted before and it's great for when we really need to use it; I love when I can stay away from it;
* tldap if it supports python3, it might be worth some investigation(seems that it can work with django but it might not depend on it which could be a big plus)

Unfortunately, I don't have the free cycles to adapt your changes to use ldap3.
Guess we'll have to wait for someone which needs python3 support(salt is paving it's way to support python3) which also has some free time to take up of the task.

jfindlay commented:
Go Go Jenkins!

cachedout commented on .gitignore:
```diff
@@ -18,14 +18,17 @@ MANIFEST
 # - /some/path$ git clone https://github.com/thatch45/salt.git
 # - /some/path$ virtualenv --python=/usr/bin/python2.6 salt
 bin/
+/etc/
```
What's the need for these absolute path ignores?

richardhansen replied:
It's to ensure that only a top-level `etc` directory is ignored, not `foo/etc`.

cachedout commented on salt/modules/ldap2.py:
```diff
@@ -0,0 +1,423 @@
+    if method is None:
+        return
+    try:
+        bindfn = globals()['_bind_' + method]
```
Is there really no other way to get this value other than inspecting globals?

richardhansen replied:
@cachedout: I wrote it this way to make it easy to add additional bind methods.  I don't really see any other bind methods being added, so I can do away with this inspection.

cachedout commented on tests/unit/states/ldap_test.py:
```diff
@@ -0,0 +1,323 @@
+This code is gross.  I started out trying to remove some of the
+duplicate code in the test cases, and before I knew it the test code
+was an ugly second implementation.
+
```
I'll take gross tests over no tests. Thank you for writing them! :+1:

cachedout commented:
Hi @richardhansen. This looks really good! I have a few really minor questions that I left as in-line comments. If you could have a look, we can probably get this merged today. Thanks again!

cachedout commented:
@richardhansen I would love to get this merged. Did you get a chance to review the comments that were left on this PR?

richardhansen commented:
Before merging this I would like to tweak the API so that it's possible to swap out the Python `ldap` module with the Python `ldap3` module without breaking users (much).

rhansen committed:
only ignore virtualenv directories at the top

To avoid collateral damage, only ignore the virtualenv directories if
they appear at the top of the working directory.

Also, only ignore lib64 if it is a directory (mostly to match the
other entries).

richardhansen commented:
I just pushed a reroll.  Changes:

  * split the `.gitignore` commit into two:  one to add the additional directories, another to change them all from `foo/` to `/foo/`
  * eliminate premature generalization in `_bind()` (avoid looking at `globals()`)
  * rename the `ldap2` execution module to `ldap3` to prevent users from thinking that it's for protocol version 2, and because eventually this will be modified to default to using the `ldap3` Python module if installed
  * change the API of the `ldap3` execution module to support both the `ldap` and `ldap3` Python modules.  The code still only supports `ldap`, but it should now be possible to add support for `ldap3` in the future without breaking users too much

Would be nice to fix before merging:

  * fix lint issues
  * throw away that gross test code and put in something better

richardhansen commented:
I just pushed another reroll.  Changes:

  * fix lint issues
  * improve exception handling.  Now all exceptions related to interacting with the LDAP server are instances of `ldap3.LDAPError` to avoid catch-all `except` blocks

The test code is still gross and incomplete, and it would be nice to support the `ldap3` Python module, but those can be fixed later.  I think it's good enough to merge now.

cachedout commented:
Thanks for the follow-up here, @richardhansen. Once the tests are finished running, assuming everything passes we will get this merged. This is great stuff! Thank you!

richardhansen commented:
About the lint errors:
  * `ldap3.py` line 15:  That line is blank; why does it say "incompatible Python 3 code found"?
  * `ldap3.py` line 55 deprecated raise error:  What's the alternative?
  * `ldap3.py` line 55 incompatible raise:  `six.raise_from()` doesn't quite do what I want, but maybe I can use it with `six.PY2` to fix this

richardhansen commented:
OK, all lint issues have been addressed and it passes all tests.

richardhansen commented:
Just pushed a reroll.  No changes other than rebase onto the latest `develop`.

Assuming the rebase still passes all tests, I think it's good enough to merge.  Any additional feedback?

richardhansen commented:
Rebased onto latest `develop`.

richardhansen commented:
Rebased onto latest `develop`.

thedrow commented:
This looks great. I'd love to have a way to use the same syntax as the user/group states for basic ldap management so:
```yaml
ldap.user.present:
  - name: myuser
  # etc
```
will add a user named myuser to the LDAP server.
I don't think that most of us know LDAP query syntax by heart.

If that's not going to happen in this PR that's OK but that means that this PR won't address #25911.

richardhansen commented:
@thedrow,

> I'd love to have a way to use the same syntax as the user/group states for basic ldap management

It shouldn't be difficult to create a new state module based on the new `ldap3` execution module that provides an easy and consistent interface for managing users and groups in an LDAP directory.  One could even extend the existing `user` and `group` state modules (perhaps via an `ldap: True` argument) to provide that functionality.  It would require site-specific configuration for stuff like server URL, bind info, RFC 2307 vs. rfc2307bis schema, etc.

> If that's not going to happen in this PR that's OK

It's not.  That's a new feature that's best suited for a new PR.  I personally have no plans to implement such a feature; the code in this PR is sufficient for my needs, and everyone should be able to use it to create a workable (but not simple) solution for their needs.

> but that means that this PR won't address #25911.

The wording of that issue is pretty open-ended.  If you're looking for a specific interface then you might want to edit the issue's description to add more details.

thedrow commented:
> I think that we can adjust our current user and group states to map directly to LDAP concepts but I'm not sure how the syntax would look like.
Just saying that we shouldn't close #25911 just yet until that kind of functionality is added.

cachedout commented:
Glad we could finally get this in. This is great work @richardhansen ! Thank you!

richardhansen commented:
Thank you for merging it!
